File: TypePolynomialModelPython/HybridMemoryPolynomialModel.py
import numpy as np
import pandas as pd
from sklearn.metrics import mean_squared_error

class HybridMemoryPolynomialModel:
    def __init__(self, df, K, M, K_e, M_e):
        self.df = self.prepare_data(df)
        self.K = K
        self.M = M
        self.K_e = K_e
        self.M_e = M_e
        self.A = None
        self.input_data, self.output_data = self.df['input'].to_numpy(), self.df['output'].to_numpy()

    def prepare_data(self, df):
        df.columns = df.columns.str.lower()
        df['input'] = df['input'].apply(lambda x: complex(x))
        df['output'] = df['output'].apply(lambda x: complex(x))
        df['input_real'] = df['input'].apply(lambda x: x.real)
        df['input_imag'] = df['input'].apply(lambda x: x.imag)
        df['output_real'] = df['output'].apply(lambda x: x.real)
        df['output_imag'] = df['output'].apply(lambda x: x.imag)
        df['input'] = df['input_real'] + 1j * df['input_imag']
        df['output'] = df['output_real'] + 1j * df['output_imag']
        return df

    def fit(self):
        phi_HMP = []
        for n in range(max(self.M, self.M_e), len(self.input_data)):
            row = self.construct_mp_terms(n) + self.construct_envmp_terms(n)
            phi_HMP.append(row)
        
        max_len = max(len(row) for row in phi_HMP)
        phi_HMP = np.array([np.pad(row, (0, max_len - len(row)), mode='constant') for row in phi_HMP])
        
        y_trimmed = self.output_data[max(self.M, self.M_e):]
        self.A, _, _, _ = np.linalg.lstsq(phi_HMP, y_trimmed, rcond=None)

    def construct_mp_terms(self, n):
        terms = []
        for m in range(self.M + 1):
            for k in range(1, self.K + 1):
                term = self.input_data[n - m] * np.abs(self.input_data[n - m])**(k - 1)
                terms.append(term)
        return terms

    def construct_envmp_terms(self, n):
        terms = []
        for m in range(1, self.M_e + 1):
            for k in range(2, self.K_e + 1):
                if n - m >= 0:
                    term = self.input_data[n] * np.abs(self.input_data[n - m])**(k - 1)
                    terms.append(term)
        return terms

    def predict(self):
        if self.A is None:
            raise ValueError("Model has not been fitted yet.")
        
        y_pred = []
        for n in range(max(self.M, self.M_e), len(self.input_data)):
            phi_n = self.construct_mp_terms(n) + self.construct_envmp_terms(n)
            phi_n = np.pad(phi_n, (0, len(self.A) - len(phi_n)), mode='constant')
            y_n = np.dot(phi_n, self.A)
            y_pred.append(y_n)
        
        return np.array(y_pred)

    def calculate_rmse(self, y_true, y_pred):
        rmse_real = np.sqrt(mean_squared_error(y_true.real, y_pred.real))
        rmse_imag = np.sqrt(mean_squared_error(y_true.imag, y_pred.imag))
        return rmse_real, rmse_imag

# Загрузка и подготовка данных
df = pd.read_csv('Amp_C_train.txt')

# Параметры модели
K = 3    # Порядок нелинейности для полинома памяти
M = 2    # Глубина памяти для полинома памяти
K_e = 3  # Порядок нелинейности для огибающей полинома памяти
M_e = 3  # Глубина памяти для огибающей полинома памяти

# Создание и обучение модели
model = HybridMemoryPolynomialModel(df, K, M, K_e, M_e)
model.fit()

# Предсказание
y_pred = model.predict()

# Расчет RMSE
y_true = df['output'].to_numpy()[max(M, M_e):]
rmse_real, rmse_imag = model.calculate_rmse(y_true, y_pred)
print(f'RMSE (Real part): {rmse_real}, RMSE (Imaginary part): {rmse_imag}')

File: TypePolynomialModelC/HybridMemoryPolynomialModel.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <complex>
#include <armadillo>
#include <cmath>
#include <stdexcept>
#include <iomanip>

class HybridMemoryPolynomialModel {
private:
    arma::cx_mat data;
    arma::cx_colvec coefficients;
    int K;    // Порядок нелинейности для полинома памяти
    int M;    // Глубина памяти для полинома памяти
    int K_e;  // Порядок нелинейности для огибающей полинома памяти
    int M_e;  // Глубина памяти для огибающей полинома памяти

public:
    HybridMemoryPolynomialModel(const std::string& filename, int K, int M, int K_e, int M_e)
        : K(K), M(M), K_e(K_e), M_e(M_e), coefficients(arma::cx_colvec()) {
        int rows;
        data = load_data(filename, rows);
        if (rows <= std::max(M, M_e)) {
            throw std::runtime_error("Error: Not enough data samples or incorrect data format.");
        }
    }

    arma::cx_mat load_data(const std::string& filename, int& rows) {
        std::ifstream file(filename);
        if (!file.is_open()) {
            throw std::runtime_error("Error: Could not open file " + filename);
        }

        std::string line;
        std::vector<std::complex<double>> inputs, outputs;

        while (std::getline(file, line)) {
            if (line.find("Time") != std::string::npos) continue;

            std::stringstream ss(line);
            std::string token;
            double real_in, imag_in, real_out, imag_out;
            int index = 0;

            while (std::getline(ss, token, ',')) {
                std::replace(token.begin(), token.end(), '(', ' ');
                std::replace(token.begin(), token.end(), ')', ' ');
                std::replace(token.begin(), token.end(), 'j', ' ');
                std::stringstream val_stream(token);
                val_stream >> real_in >> imag_in;

                if (index == 1) {
                    inputs.emplace_back(real_in, imag_in);
                } else if (index == 2) {
                    outputs.emplace_back(real_in, imag_in);
                }
                ++index;
            }
        }

        rows = inputs.size();
        arma::cx_mat data(inputs.size(), 2);
        for (size_t i = 0; i < inputs.size(); ++i) {
            data(i, 0) = inputs[i];
            data(i, 1) = outputs[i];
        }

        return data;
    }

    void fit() {
        int n_samples = data.n_rows - std::max(M, M_e);
        int max_terms = (M + 1) * K + M_e * (K_e - 1);
        arma::cx_mat phi_HMP(n_samples, max_terms, arma::fill::zeros);
        arma::cx_colvec y_trimmed(n_samples);

        for (int n = std::max(M, M_e); n < data.n_rows; ++n) {
            std::vector<std::complex<long double>> row;
            construct_mp_terms(row, n);
            construct_envmp_terms(row, n);

            for (size_t i = 0; i < row.size(); ++i) {
                phi_HMP(n - std::max(M, M_e), i) = row[i];
            }

            y_trimmed(n - std::max(M, M_e)) = data(n, 1);
        }

        coefficients = arma::solve(phi_HMP, y_trimmed);
    }

    void construct_mp_terms(std::vector<std::complex<long double>>& terms, int n) {
        for (int m = 0; m <= M; ++m) {
            for (int k = 1; k <= K; ++k) {
                std::complex<long double> term = data(n - m, 0) * std::pow(std::abs(data(n - m, 0)), k - 1);
                terms.push_back(term);
            }
        }
    }

    void construct_envmp_terms(std::vector<std::complex<long double>>& terms, int n) {
        for (int m = 1; m <= M_e; ++m) {
            for (int k = 2; k <= K_e; ++k) {
                if (n - m >= 0) {
                    std::complex<long double> term = data(n, 0) * std::pow(std::abs(data(n - m, 0)), k - 1);
                    terms.push_back(term);
                }
            }
        }
    }

    arma::cx_colvec predict() {
        if (coefficients.n_elem == 0) {
            throw std::runtime_error("Model has not been fitted yet.");
        }

        int n_samples = data.n_rows - std::max(M, M_e);
        arma::cx_colvec y_pred(n_samples);

        for (int n = std::max(M, M_e); n < data.n_rows; ++n) {
            std::vector<std::complex<long double>> phi_n;
            construct_mp_terms(phi_n, n);
            construct_envmp_terms(phi_n, n);

            arma::cx_rowvec phi_vec(phi_n.size());
            for (size_t i = 0; i < phi_n.size(); ++i) {
                phi_vec(i) = phi_n[i];
            }

            y_pred(n - std::max(M, M_e)) = arma::dot(phi_vec, coefficients);
        }

        return y_pred;
    }

    std::pair<long double, long double> calculate_rmse(const arma::cx_colvec& y_true, const arma::cx_colvec& y_pred) {
        long double rmse_real = arma::norm(arma::real(y_true) - arma::real(y_pred)) / std::sqrt(y_true.size());
        long double rmse_imag = arma::norm(arma::imag(y_true) - arma::imag(y_pred)) / std::sqrt(y_true.size());

        return {rmse_real, rmse_imag};
    }

    arma::cx_colvec get_output_data() const {
        return data.col(1).tail(data.n_rows - std::max(M, M_e));
    }
};

int main(int argc, char* argv[]) {
    try {
        if (argc != 6) {
            std::cerr << "Usage: " << argv[0] << " <filename> <K> <M> <K_e> <M_e>" << std::endl;
            return 1;
        }

        std::string filename = argv[1];
        int K = std::stoi(argv[2]);
        int M = std::stoi(argv[3]);
        int K_e = std::stoi(argv[4]);
        int M_e = std::stoi(argv[5]);

        HybridMemoryPolynomialModel model(filename, K, M, K_e, M_e);
        model.fit();

        arma::cx_colvec y_pred = model.predict();
        arma::cx_colvec y_true = model.get_output_data();

        auto [rmse_real, rmse_imag] = model.calculate_rmse(y_true, y_pred);

        std::cout << std::fixed << std::setprecision(6);
        std::cout << "RMSE (Real part): " << rmse_real << ", RMSE (Imaginary part): " << rmse_imag << std::endl;

    } catch (const std::exception& e) {
        std::cerr << e.what() << std::endl;
        return 1;
    }

    return 0;
}
